### PR TITLE
Fix MultiNodeWeightedSampler tests

### PR DESCRIPTION
**Pull request by ramanishsingh**
Due to the stochastic nature of the MultiNodeWeightedSampler and the way StopIteration is called when multiple datasets are involved, some test results are dependent on the OS/seed. Trying to fix seed and change testing conditions with this PR.

### CODE REVIEW

File: test/nodes/test_multi_node_weighted_sampler.py [hunk 1]
```diff
@@ -24,7 +24,7 @@ def setUp(self) -> None:
         self._num_datasets = 4
         self._weights_fn = lambda i: 0.1 * (i + 1)
         self._num_epochs = 3
-
+        self._seed = 42
```
Comment (divyanshk) [inline test/nodes/test_multi_node_weighted_sampler.py:27]: Should we make the test loop over multiple seeds?

File: test/nodes/test_multi_node_weighted_sampler.py [hunk 2]
```diff
@@ -82,20 +81,21 @@ def test_multi_node_weighted_batch_sampler_zero_weights(
...
         mixer = self._setup_multi_node_weighted_sampler(
             self._num_samples,
             self._num_datasets,
             self._weights_fn,
             stop_criteria=StopCriteria.FIRST_DATASET_EXHAUSTED,
+            seed=seed,
         )
 
-        for _ in range(self._num_epochs):
+        for _ in range(3):
```
Comment (divyanshk) [inline test/nodes/test_multi_node_weighted_sampler.py:98]: Should be `self._num_epochs`

File: test/nodes/test_multi_node_weighted_sampler.py [hunk 3]
```diff
@@ -104,17 +104,19 @@ def test_multi_node_weighted_sampler_first_exhausted(self) -> None:
             # Check max item count for dataset is exactly _num_samples
             self.assertEqual(max(dataset_counts_in_results), self._num_samples)
 
-            # Check only one dataset has been exhausted
-            self.assertEqual(dataset_counts_in_results.count(self._num_samples), 1)
+            # Check that max items are taken from at least one dataset
+            self.assertGreaterEqual(dataset_counts_in_results.count(self._num_samples), 1)
```
Comment (divyanshk) [inline test/nodes/test_multi_node_weighted_sampler.py:108]: We can also add a check that no dataset has count more than `self._num_samples`.

Comment (ramanishsingh) [reply test/nodes/test_multi_node_weighted_sampler.py:108]: I think we do not need that check because we have `self.assertEqual(max(dataset_counts_in_results), self._num_samples)` in the prevous line
so that ensures no dataset item count can be greater than self._num_samples

File: test/nodes/test_multi_node_weighted_sampler.py (hunk 2 above)
Comment (ramanishsingh) [reply test/nodes/test_multi_node_weighted_sampler.py:98]: Fixed!

File: test/nodes/test_multi_node_weighted_sampler.py (hunk 3 above)
Comment (divyanshk) [reply test/nodes/test_multi_node_weighted_sampler.py:108]: Ah right!

File: test/nodes/test_multi_node_weighted_sampler.py (hunk 1 above)
Comment (divyanshk) [reply test/nodes/test_multi_node_weighted_sampler.py:27]: Synced offline, we can probably remove the _seed and add a range along with parameterized.expand to test multiple seeds for relevant unit tests.

File: test/nodes/test_multi_node_weighted_sampler.py [hunk 4]
```diff
@@ -86,13 +87,15 @@ def test_multi_node_weighted_batch_sampler_zero_weights(
                 weights={f"ds{i}": 10 * i for i in range(self._num_datasets)},
             )
 
-    def test_multi_node_weighted_sampler_first_exhausted(self) -> None:
+    @parameterized.expand(range(10))
+    def test_multi_node_weighted_sampler_first_exhausted(self, seed) -> None:
         """Test MultiNodeWeightedSampler with stop criteria FIRST_DATASET_EXHAUSTED"""
         mixer = self._setup_multi_node_weighted_sampler(
```
Comment (divyanshk) [inline test/nodes/test_multi_node_weighted_sampler.py:93]: one tiny nit: can you make this `node = self._setup_multi_node_weighted_sampler(` here as well as in other places we call _setup_multi_node_weighted_sampler, thanks!

**Review by divyanshk — APPROVED** [review]
LGTM! Thanks for the fix!
